nitorch/tools/affine_reg.py

```diff
@@ -0,0 +1,961 @@
+    dtype = mats[0].dtype
+    device = mats[0].device
+    N = len(mats)
+    Mat = torch.zeros((4, 4, N), dtype=dtype, device=device)
+    Dim = torch.zeros((3, N), dtype=dtype, device=device)
+    for i in range(N):
+        Mat[..., i] = mats[i].clone()
+        Dim[..., i] = dims[i].clone()
+    # Compute mean-space
+    dim, mat, _ = mean_space(Mat, Dim)
+    dim = tuple(dim.cpu().int().tolist())
+    mat = mat.type(dtype)
+
+    return mat, dim
+
+
+def _hist_2d(img0, img1, fwhm=7, mx_int=255):
+    """Make 2D histogram.
+
+    Pre-requisites:
+    * Images same size
+    * Images same min and max intensities
+
+    # Naive method for computing a 2D histogram
+    h = torch.zeros((mx_int + 1, mx_int + 1))
+    for n in range(num_vox):
+        h[img0[n], mg1[n]] += 1
+
+    Parameters
+    ----------
+    img0 : (X, Y, Z) tensor_like
+        First image volume.
+    img1 : (X, Y, Z) tensor_like
+        Second image volume.
+    fwhm : float, default=7
+        Full-width at half max of Gaussian kernel, for smoothing
+        histogram.
+    mx_int : float, default=255
+        Max intensity of input images.
+
+    Returns
+    ----------
+    H : (mx_int + 1, mx_int + 1) tensor_like
+        Joint intensity histogram.
+
+    """
+    fwhm = (fwhm,) * 2
+
+    # Convert each 'coordinate' of intensities to an index
+    # (replicates the sub2ind function of MATLAB)
+    img0 = img0.flatten().floor()
+    img1 = img1.flatten().floor()
+    sub = torch.stack((img0, img1),
+                      dim=1)  # (num_vox, 2)
+    to_ind = torch.tensor((1, mx_int + 1),
+                          dtype=sub.dtype, device=img0.device)[..., None]  # (2, 1)
+    ind = torch.tensordot(sub, to_ind, dims=([1], [0]))  # (nvox, 1)
+
+    # Build histogram H by adding up counts according to the indicies in ind
+    H = torch.zeros(mx_int + 1, mx_int + 1, device=img0.device, dtype=ind.dtype)
+    H.put_(ind.long(), torch.ones(1, device=img0.device, dtype=ind.dtype).expand_as(ind), accumulate=True)
+
+    # Smooth the histogram
+    smo = smooth(('gauss',) * 2, fwhm=fwhm, device=img0.device, dtype=torch.float32, sep=True)
+    # Padding amount for subsequent convolution
+    size_pad = (smo[0].shape[2], smo[1].shape[3])
+    size_pad = (torch.tensor(size_pad) - 1) // 2
+    size_pad = tuple(size_pad.int().cpu().tolist())
+    # Smooth deformation with Gaussian kernel (by convolving)
+    H = pad(H, size_pad, side='both')
+    H = H[None, None, ...]
+    H = F.conv1d(H, smo[0])
+    H = F.conv1d(H, smo[1])[0, 0, ...]
+
+    # Add eps
+    H = H + 1e-7
+
+    # # Visualise histogram
+    # plt.figure(num=fig_num)
+    # plt.imshow(H,
+    #     cmap='coolwarm', interpolation='nearest',
+    #     aspect='equal', vmax=0.05*H.max())
+    # plt.axis('off')
+    # plt.show()
+
+    return H
+
+
+def _init_torch(opt):
+    """Check device and set some Torch stuff.
+    """
+    opt['device'] = torch.device(opt['device'] if torch.cuda.is_available() else 'cpu')
```

Comment:
We have to be careful with this:
- I might have a GPU but know that my images are too big so want to run the code on CPU anyway
- What if I run some DL stuff afterward: do I fuck everything up by having`set_grad_enabled(False)`.
- I have looked a bit what the benchmarking does, and it is only used by the convolution functions, and is only useful if you perform a lot of convolutions with the same input size (https://discuss.pytorch.org/t/what-does-torch-backends-cudnn-benchmark-do/5936/2). It is probably useful here because the joint histogram always have the same size. Do we need to set it back to its default value afterward? Should we use a `with` statement?

Another solution could be to wrap the main code in a `with torch.no_grad():` statement

Reply:
Good point, I have removed the '_init_torch' function.

How would I go about wrapping the code like that, in a nice way?

Reply:
It's really focused on gradient-based optimization, but the base class is there: https://pytorch.org/docs/stable/optim.html#torch.optim.Optimizer

Maybe it's not suited to powell. I don't know.
Also, I wonder if we could use autograd to optimise the whole thing with pytorch's optimizers.
Last thing we could check is how things are done in BoTorch (https://botorch.org/docs/design_philosophy)

Reply:
The quick/dirty way I see is to have the main code in a subfunction (_run_affine), and then wrap it in the actual api function:

```{python}
def run_affine(...):
    with torch.no_grad():
        with device:
            _run_affine(...)
```

The less ugly way is to write our own class that implements `__enter__` and `__exit__`. See https://www.python.org/dev/peps/pep-0343/ and https://docs.python.org/3/reference/compound_stmts.html#with

Reply:
Okay, I just added 'with torch.no_grad():' at the start of run_affine, probably will do for now :)